### PR TITLE
ddl: Batch backfilling index and column data

### PR DESCRIPTION
**Pull request by zimulala**
Using 1 TiDB 1 TiKV, total rows of test table is 336253.
Before:
alter add index takes 14 min 36.58 sec
alter add column int(1) not null takes 14 min 22.27 sec
After batch=100:
alter add index takes 3 min 14.59 sec
alter add column int(1) not null takes 1 min 56.58 sec
After batch=128:
alter add index takes 2 min 54.49 sec
alter add column int(1) not null takes 1 min 16.95 sec

### CODE REVIEW

**Comment by zimulala** [comment]
PTAL @coocood @disksing

File: ddl/column.go [hunk 1]
```diff
@@ -334,11 +334,54 @@ func (d *ddl) backfillColumn(t table.Table, columnInfo *model.ColumnInfo, reorgI
...
+		}
+		if _, ok := rowColumns[colID]; ok {
+			// The column is already added by update or insert statement, skip it.
+			continue
+		}
+
+		newColumnIDs := make([]int64, 0, len(rowColumns)+1)
+		newRow := make([]types.Datum, 0, len(rowColumns)+1)
+		for colID, val := range rowColumns {
+			newColumnIDs = append(newColumnIDs, colID)
```
Comment (hanfei1991) [inline ddl/column.go]: It's better to use copy

File: ddl/column.go [hunk 2]
```diff
@@ -334,11 +334,54 @@ func (d *ddl) backfillColumn(t table.Table, columnInfo *model.ColumnInfo, reorgI
...
-func (d *ddl) backfillColumnData(t table.Table, columnInfo *model.ColumnInfo, handles []int64, reorgInfo *reorgInfo) error {
-	var (
-		defaultVal types.Datum
-		err        error
-	)
+func (d *ddl) batchBackfillColumn(t table.Table, colID int64, handles []int64, colMap map[int64]*types.FieldType,
+	defaultVal types.Datum, txn kv.Transaction) (int64, error) {
+	nextHandle := handles[0]
+	for _, handle := range handles {
+		log.Debug("[ddl] backfill column...", handle)
```
Comment (hanfei1991) [inline ddl/column.go]: Can we remove the debug log here?

File: ddl/column.go (hunk 1 above)
Comment (zimulala) [inline ddl/column.go]: Why ?

File: ddl/index.go [hunk 3]
```diff
@@ -452,51 +452,67 @@ func (d *ddl) getSnapshotRows(t table.Table, version uint64, seekHandle int64) (
...
-			if terror.ErrorEqual(err1, kv.ErrNotExist) {
-				// row doesn't exist, skip it.
-				return nil
-			}
-			if err1 != nil {
-				return errors.Trace(err1)
-			}
-
-			exist, _, err1 := kvX.Exist(txn, vals, handle)
+			nextHandle, err1 := d.batchBackfillIndex(t, kvIdx, handles[:endIdx], txn)
```
Comment (hanfei1991) [inline ddl/index.go]: why not use err ?

File: ddl/index.go [hunk 4]
```diff
@@ -452,51 +452,67 @@ func (d *ddl) getSnapshotRows(t table.Table, version uint64, seekHandle int64) (
...
+		nextHandle = handle
+	}
+	return nextHandle, nil
+}
+
+func (d *ddl) backfillTableIndex(t table.Table, indexInfo *model.IndexInfo, handles []int64, reorgInfo *reorgInfo) error {
+	var endIdx int
+	kvIdx := tables.NewIndex(t.Meta(), indexInfo)
+	for len(handles) > 0 {
+		if len(handles) >= defaultBackfillBatchSize {
```
Comment (hanfei1991) [inline ddl/index.go]: How about use > insead of >=

File: ddl/column.go [hunk 5]
```diff
@@ -347,55 +390,36 @@ func (d *ddl) backfillColumnData(t table.Table, columnInfo *model.ColumnInfo, ha
...
+
+	var endIdx int
+	for len(handles) > 0 {
+		if len(handles) >= defaultBackfillBatchSize {
+			endIdx = defaultBackfillBatchSize
+		} else {
+			endIdx = len(handles)
+		}
+
+		err = kv.RunInNewTxn(d.store, true, func(txn kv.Transaction) error {
```
Comment (hanfei1991) [inline ddl/column.go]: err := .... is better?

File: ddl/column.go [hunk 6]
```diff
@@ -334,11 +334,54 @@ func (d *ddl) backfillColumn(t table.Table, columnInfo *model.ColumnInfo, reorgI
...
+		if err != nil {
+			return 0, errors.Trace(err)
+		}
+		err = txn.Set(rowKey, newRowVal)
+		if err != nil {
+			return 0, errors.Trace(err)
+		}
+	}
+
+	return nextHandle, nil
```
Comment (hanfei1991) [inline ddl/column.go]: Where do you update nextHandle?

Comment (zimulala) [inline ddl/column.go]: The nextHandle is In the function of backfillColumn.

File: ddl/index.go (hunk 4 above)
Comment (zimulala) [inline ddl/index.go]: I think it doesn't necessary.

**Comment by hanfei1991** [comment]
LGTM

**Comment by zimulala** [comment]
PTAL @coocood @disksing @tiancaiamao @shenli

**Comment by coocood** [comment]
LGTM

**Pull request merged**